cocotb/handle.py

```diff
@@ -625,7 +655,12 @@ def setimmediatevalue(self, value):
             self._log.critical("Unsupported type for value assignment: %s (%s)", type(value), repr(value))
             raise TypeError("Unable to set simulator value with type %s" % (type(value)))
 
-        simulator.set_signal_val_binstr(self._handle, value.binstr)
+        simulator.set_signal_val_binstr(self._handle, set_action, value.binstr)
+
+    def check_for_set_action(self, value):
+        if not isinstance(value, _SetAction):
+            value = Deposit(value)
```

Comment:
It _might_ be worth making this `return value, 0  # GPI_DEPOSIT`. It would be faster, and is possibly the most likely case. However, it is less readable.

Another possibly-faster spelling would be:
```
try:
    as_gpi_args_for = value._as_gpi_args_for
except AttributeError:
    return value, 0  # GPI_DEPOSIT
else:
    return as_gpi_args_for(self)
```

Difficult to know without profiling.

Reply:
The first one makes sense to me, at it is (for most testbenches) the most common occurrence. Also it's still pretty readable, given that the return values are clearly named in the callee and the scope is pretty limited. The try/except one tends to look a bit messy, though.

Reply:
The idea was that `try / except` may be faster than calling `isinstance`. I don't have much to back that up though.

Reply:
I built a quick out-of-context benchmark to check. It appears the try-except version fares much worse whenever an exception needs to be handled (i.e. for the most common code path in this situation).

https://gist.github.com/thasti/a2c4e553b39a9d914b34766eb42fb779

The less common path (a SetAction instance is passed) is pretty much the same cost for both tests, but the more common one is faster in the "isinstance" implementation:

```
Try/Except - Except: 0.385
Try/Except - Else:   0.208
isinstance - No:     0.125
isinstance - Yes:    0.194
```

Also (not in the gist): Directly returning the tuple instead of constructing yet another object saves considerable overhead for such small data structures, so the current version appears to be the best in performance evaluated so far.

Reply:
```
Try/Except - Except: 0.407
Try/Except - Else:   0.207
isinstance - No:     0.238
isinstance - Yes:    0.201
getattr - None:      0.133
getattr - not None:  0.232
```

Quite a bit faster for the most common path, while not slower for the less common one. Changed in the PR.

Reply:
Actually I bodged this last benchmark by not reverting the modification to test2 that creates a new instance. So, when run as in the gist, the results are:

```
Try/Except - Except: 0.383
Try/Except - Else:   0.201
isinstance - No:     0.101
isinstance - Yes:    0.194
getattr - None:      0.125
getattr - not None:  0.221
```

And therefore the isinstance-based one is consistently the fastest. Reverted the code changes to use this implementation.